src/quantization/sq4_quantizer.h

```diff
@@ -0,0 +1,194 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <cstring>
+#include <limits>
+#include <vector>
+
+#include "quantizer.h"
+#include "simd/sq4_simd.h"
+
+namespace vsag {
+
+template <MetricType metric = MetricType::METRIC_TYPE_L2SQR>
+class SQ4Quantizer : public Quantizer<SQ4Quantizer<metric>> {
+public:
+    explicit SQ4Quantizer(int dim);
+
+    bool
+    TrainImpl(const DataType* data, uint64_t count);
+
+    bool
+    EncodeOneImpl(const DataType* data, uint8_t* codes) const;
+
+    bool
+    EncodeBatchImpl(const DataType* data, uint8_t* codes, uint64_t count);
+
+    bool
+    DecodeOneImpl(const uint8_t* codes, DataType* data);
+
+    bool
+    DecodeBatchImpl(const uint8_t* codes, DataType* data, uint64_t count);
+
+    inline float
+    ComputeImpl(const uint8_t* codes1, const uint8_t* codes2);
+
+    inline void
+    ProcessQueryImpl(const DataType* query, Computer<SQ4Quantizer>& computer) const;
+
+    inline void
+    ComputeDistImpl(Computer<SQ4Quantizer>& computer, const uint8_t* codes, float* dists) const;
+
+private:
+    std::vector<DataType> lower_bound_;
+    std::vector<DataType> diff_;
+};
+
+template <MetricType metric>
+SQ4Quantizer<metric>::SQ4Quantizer(int dim) : Quantizer<SQ4Quantizer<metric>>(dim) {
+    this->code_size_ = (dim + 1) >> 1 << 1;
+    lower_bound_.resize(dim, std::numeric_limits<DataType>::max());
+    diff_.resize(dim, std::numeric_limits<DataType>::min());
+}
+
+template <MetricType metric>
+bool
+SQ4Quantizer<metric>::TrainImpl(const DataType* data, uint64_t count) {
+    if (data == nullptr) {
+        return false;
+    }
+
+    for (uint32_t i = 0; i < count; i++) {
+        for (uint32_t d = 0; d < this->dim_; d++) {
+            auto val = data[i * this->dim_ + d];
+            if (val > diff_[d]) {
+                diff_[d] = val;
+            }
+            if (val < lower_bound_[d]) {
+                lower_bound_[d] = val;
+            }
+        }
+    }
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
```

Comment:
dim_ is uint64_t type, d is uint32_t type, we should avoid implicit conversions, please make it unify